[BUG] libsinsp-example crash on arm64 + modern_bpf

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks! -->

**Describe the bug**

Running the `sinsp-example` test program on an arm64 host with the modern_bpf driver causes a seg fault. Looking at it with a debugger it seems that the root cause is related to malformed events a the scap level and it is really likely a modern_bpf driver problem.

**How to reproduce it**

1. Patch the `sinsp-example` program removing the output (for some reason this step is essential to get the repro)
```
diff --git a/userspace/libsinsp/examples/test.cpp b/userspace/libsinsp/examples/test.cpp
index d0334b4bd..f7fc55c2b 100644
--- a/userspace/libsinsp/examples/test.cpp
+++ b/userspace/libsinsp/examples/test.cpp
@@ -415,11 +415,12 @@ int main(int argc, char** argv) {
                        cout << "[ERROR] " << error_msg << endl;
                });
                if(ev != nullptr) {
-                       sinsp_threadinfo* thread = ev->get_thread_info();
-                       if(!thread || g_all_threads || thread->is_main_thread()) {
-                               dump(inspector, ev);
-                               num_events++;
-                       }
+                       //sinsp_threadinfo* thread = ev->get_thread_info();
+                       //if(!thread || g_all_threads || thread->is_main_thread()) {
+                       //      dump(inspector, ev);
+                       //      num_events++;
+                       //}
                }
        }
        std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
``` 
2. Run `sinsp-example -m`
3. Once the capture is started run `stress-ng --exec 1 --open 1` and the crash should appear almost immediately.

**Expected behaviour**

There should not be any malformed event.

**Screenshots**


<!-- If applicable, add screenshots to help explain your problem. -->

**Environment**

<!-- Please complete the following info. -->

- Falco version: master
- System info:
```
{
  "machine": "aarch64",
  "nodename": "redacted",
  "release": "6.5.0-1022-aws",
  "sysname": "Linux",
  "version": "#22~22.04.1-Ubuntu SMP Fri Jun 14 19:23:09 UTC 2024"
}

```
- Cloud provider or hardware configuration: ec2 arm64 machine
- OS:
```
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```
- Kernel:
```
Linux redacted 6.5.0-1022-aws #22~22.04.1-Ubuntu SMP Fri Jun 14 19:23:09 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux
```
- Installation method:
source

**Additional context**

<!-- Add any other context about the problem here. -->

--- a/userspace/libpman/src/ringbuffer_definitions.h
+++ b/userspace/libpman/src/ringbuffer_definitions.h
@@ -20,6 +20,7 @@ limitations under the License.
 
 #include "state.h"
 #include <linux/bpf.h>
+#include <libscap/scap_barrier.h>
 
 /* Taken from libbpf: /src/ringbuf.c */
 struct ring {
@@ -50,23 +51,3 @@ static inline int roundup_len(uint32_t len) {
 	/* round up to 8 byte alignment */
 	return (len + 7) / 8 * 8;
 }
-
-/* Taken from libbpf: `include/linux/compiler.h` */
-
-#define READ_ONCE(x) (*(volatile typeof(x) *)&x)
-#define WRITE_ONCE(x, v) (*(volatile typeof(x) *)&x) = (v)
-
-#define barrier() asm volatile("" ::: "memory")
-
-#define smp_store_release(p, v) \
-	do {                        \
-		barrier();              \
-		WRITE_ONCE(*p, v);      \
-	} while(0)
-
-#define smp_load_acquire(p)              \
-	({                                   \
-		typeof(*p) ___p = READ_ONCE(*p); \
-		barrier();                       \
-		___p;                            \
-	})

--- a/userspace/libscap/engine/bpf/scap_bpf.h
+++ b/userspace/libscap/engine/bpf/scap_bpf.h
@@ -20,6 +20,7 @@ limitations under the License.
 
 #include <libscap/compat/bpf.h>
 #include <libscap/compat/perf_event.h>
+#include <libscap/scap_barrier.h>
 
 struct perf_event_sample {
 	struct perf_event_header header;
@@ -49,9 +50,7 @@ static inline void scap_bpf_get_buf_pointers(scap_device *dev,
 	*phead = header->data_head;
 	*ptail = header->data_tail;
 
-	// clang-format off
-	asm volatile("" ::: "memory");
-	// clang-format on
+	mem_barrier();
 
 	uint64_t cons = *ptail % header->data_size;  // consumer position
 	uint64_t prod = *phead % header->data_size;  // producer position
@@ -154,9 +153,7 @@ static inline void scap_bpf_advance_tail(struct scap_device *dev) {
 
 	header = (struct perf_event_mmap_page *)dev->m_buffer;
 
-	// clang-format off
-	asm volatile("" ::: "memory");
-	// clang-format on
+	mem_barrier();
 
 	ASSERT(dev->m_lastreadsize > 0);
 	/* `header->data_tail` is the consumer position. */

--- a/userspace/libscap/linux/barrier.h
+++ b/userspace/libscap/linux/barrier.h
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: Apache-2.0
 /*
-Copyright (C) 2023 The Falco Authors.
+Copyright (C) 2024 The Falco Authors.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
@@ -17,4 +17,61 @@ limitations under the License.
 */
 #pragma once
 
-#define mem_barrier() __sync_synchronize()
+// This is taken from kernel headers `/include/linux/compiler.h`
+// Used by libpman and scap_bpf engine
+
+#define READ_ONCE(x) (*(volatile typeof(x) *)&x)
+#define WRITE_ONCE(x, v) (*(volatile typeof(x) *)&x) = (v)
+
+#define barrier() asm volatile("" ::: "memory")
+
+#if defined(__x86_64__)
+
+#define smp_mb() asm volatile("lock; addl $0,-132(%%rsp)" ::: "memory", "cc")
+
+#define smp_store_release(p, v) \
+	do {                        \
+		barrier();              \
+		WRITE_ONCE(*p, v);      \
+	} while(0)
+
+#define smp_load_acquire(p)              \
+	({                                   \
+		typeof(*p) ___p = READ_ONCE(*p); \
+		barrier();                       \
+		___p;                            \
+	})
+
+#elif defined(__aarch64__)
+
+#define smp_mb() asm volatile("dmb ish" ::: "memory")
+
+#endif
+
+#ifndef smp_mb
+#define smp_mb() __sync_synchronize()
+#endif
+
+#ifndef smp_store_release
+#define smp_store_release(p, v) \
+	do {                        \
+		smp_mb();               \
+		WRITE_ONCE(*p, v);      \
+	} while(0)
+#endif
+
+#ifndef smp_load_acquire
+#define smp_load_acquire(p)              \
+	({                                   \
+		typeof(*p) ___p = READ_ONCE(*p); \
+		smp_mb();                        \
+		___p;                            \
+	})
+#endif
+
+// This is defined by us
+#if defined(__x86_64__)
+#define mem_barrier() barrier()
+#else
+#define mem_barrier() smp_mb()
+#endif

